### ISSUE TITLE
CST and AST

### ISSUE THREAD

**Issue by atahanyorganci** [atahanyorganci/alloy#24]
Concrete Syntax Tree (CST) and Abstract Syntax Tree (AST) are closely related structures and having different parser that produce either one but not the other is wasted effort.

Since CST includes info such as delimiters and whitespace characters used a CST can be cast down to AST.﻿

Extra info present in CST can be used to format source files (i.e. `cargo fmt`) or programatically produce documents from source files.

**Comment by atahanyorganci** [comment]
Procedural macros can be used to derive ASTs from CSTs a derive macro can be used to mark fields that represent white space so that they are not included in the AST. Further, span information is also redundant so `Spanned<T>` can be transformed into `T`. For self referential types which includes nearly all CST nodes `T` has to transformed into `U` where `U: AST<T>`.

```rust
#[derive(AST)]
struct BinaryExpressionCST {
    lhs: Spanned<Box<ExpressionCST>>,
    #[space]
    lw: SpannedSpace,
    op: Spanned<BinaryOperator>,
    #[space]
    rw: SpannedSpace,
    rhs: Spanned<Box<ExpressionCST>>,
}
```
Would result in

```rust
struct BinaryExpression {
    lhs: Box<Expression>,
    op: BinaryOperator,
    rhs: Box<Expression>,
}
```

For making CST to AST easy `From<CST>` can be implemented for `AST`.

```rust
impl From<CST> for AST {
    fn from(cst: &CST) -> Self {
          // destructure CST and assign to AST fields
    }
}
```

**Comment by atahanyorganci** [comment]
Definitions
- **Space field** field with `#[space]` attribute
- **Simple field** field with type that doesn't include self reference.
- **CST field** field with self reference.

Since procedural macros operate on syntax level where type information isn't present distinguishing between *simple field* and *cst field* has to be done at a syntax level as well.

**Comment by atahanyorganci** [comment]
Essentially deriving AST from CST requires following steps.

- [ ] Creating a new identifier for AST struct
- [ ] Dropping fields marked with `#[space]`
- [ ] Mapping simple fields such as `Spanned<T>` into `T`
- [ ] Mapping complex fields such as `Spanned<T>` to `U` where `U: AST<T>`

**Comment by atahanyorganci** [comment]
Current AST implementation in Alloy includes enums such as `Expression` that is discriminated union of `BinaryExpression`, `UnaryExpression` etc.

```rust
// src/ast/expression/mod.rs:21
pub enum Expression {
    Value(Value),
    Binary(BinaryExpression),
    Unary(UnaryExpression),
    Identifier(IdentifierExpression),
}
```

This AST enum has to be derived from CST one candidate CST that can be used to generate `Expression` AST is provided below.

```rust
#[derive(AST)]
pub enum ExpressionCST {
    Value(Spanned<Value>),
    Binary(BinaryExpressionCST),
    Unary(UnaryExpressionCST),
    Identifier(IdentifierExpressionCST),
}
```

**Comment by atahanyorganci** [comment]
#27 tracks implementation

**Comment by atahanyorganci** [comment]
Implemented in #37.
